src/frequenz/client/dispatch/__main__.py

```diff
@@ -0,0 +1,255 @@
+# License: All rights reserved
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""CLI and interactive client for the dispatch service."""
+
+import asyncio
+import os
+import sys
+from pprint import pformat
+from typing import Any, List
+
+import asyncclick as click
+import grpc
+from prompt_toolkit import PromptSession
+from prompt_toolkit.completion import NestedCompleter
+from prompt_toolkit.history import FileHistory
+from prompt_toolkit.patch_stdout import patch_stdout
+from prompt_toolkit.shortcuts import CompleteStyle
+
+from ._cli_types import FuzzyDateTime, FuzzyIntRange, FuzzyTimeDelta, SelectorParamType
+from ._client import Client
+
+DEFAULT_DISPATCH_API_HOST = "88.99.25.81"
+DEFAULT_DISPATCH_API_PORT = 50051
+
+
+def get_client(host: str, port: int) -> Client:
+    """Get a new client instance.
+
+    Args:
+        host: The host of the dispatch service.
+        port: The port of the dispatch service.
+
+    Returns:
+        Client: A new client instance.
+    """
+    channel = grpc.aio.insecure_channel(f"{host}:{port}")
+    return Client(channel, f"{host}:{port}")
+
+
+# Click command groups
+@click.group()
+@click.option(
+    "--host",
+    default=DEFAULT_DISPATCH_API_HOST,
+    help="Dispatch API host",
+    envvar="DISPATCH_API_HOST",
+    show_envvar=True,
+    show_default=True,
+)
+@click.option(
+    "--port",
+    default=DEFAULT_DISPATCH_API_PORT,
+    help="Dispatch API port",
+    envvar="DISPATCH_API_PORT",
+    show_envvar=True,
+    show_default=True,
+)
+@click.pass_context
+async def cli(ctx: click.Context, host: str, port: int) -> None:
+    """Dispatch Service CLI."""
+    ctx.ensure_object(dict)
+    ctx.obj["client"] = get_client(host, port)
+
+
+@cli.command("list")
+@click.pass_context
+@click.argument("microgrid-id", required=True, type=int)
+@click.option("--selector", "-s", type=SelectorParamType(), multiple=True)
+@click.option("--start-from", type=FuzzyDateTime())
+@click.option("--start-to", type=FuzzyDateTime())
+@click.option("--end-from", type=FuzzyDateTime())
+@click.option("--end-to", type=FuzzyDateTime())
+@click.option("--active", type=bool)
+@click.option("--dry-run", type=bool)
+async def list_(ctx: click.Context, /, **filters: Any) -> None:
```

Comment:
Out of curiosity, where are the keywords arguments coming from?

Reply:
you mean "who" is filling out the `filters` parameter?
Basically all the decorators for the function pre-process it before calling.
For that reason we also have to use `await cli.main(args=params, standalone_mode=False)` to invoke it properly
There are other ways to invoke the functions like `Context.invoke(..)` but there were some issues with those methods when I tried them, so the cli.main way was the most stable one.